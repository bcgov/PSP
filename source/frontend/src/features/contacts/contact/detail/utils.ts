import { ContactInfoField } from 'features/contacts/interfaces';
import { Dictionary } from 'interfaces/Dictionary';
import { IContactAddress, IContactMethod } from 'interfaces/IContact';

import { AddressTypes } from './../../../../constants/addressTypes';

// the order of this array corresponds to the expected display order
const addressSortOrder = [AddressTypes.Mailing, AddressTypes.Residential, AddressTypes.Billing];

export const sortAddresses = (a1: IContactAddress, a2: IContactAddress) => {
  if (a2.addressType.id === a1.addressType.id) {
    return 0;
  }
  const a2Index = addressSortOrder.indexOf(a2.addressType.id as AddressTypes);
  if (a2Index !== -1 && a2Index < addressSortOrder.indexOf(a1.addressType.id as AddressTypes)) {
    return 1;
  }

  return -1;
};

export function getContactInfo(
  validTypes: Dictionary<string>,
  contactMethods?: IContactMethod[],
): ContactInfoField[] {
  if (contactMethods === undefined) {
    return [];
  }
  // Get only the valid types
  let filteredFields = contactMethods.reduce(
    (accumulator: ContactInfoField[], method: IContactMethod) => {
      if (Object.keys(validTypes).includes(method.contactMethodType.id)) {
        accumulator.push({
          info: method.value,
          label: validTypes[method.contactMethodType.id],
        });
      }
      return accumulator;
    },
    [],
  );

  // Sort according to the dictionary order
  return filteredFields.sort((a, b) => {
    return Object.values(validTypes).indexOf(a.label) - Object.values(validTypes).indexOf(b.label);
  });
}

export const fakeAddresses = [
  {
    id: 1,
    rowVersion: 3,
    addressType: {
      id: 'BILLING',
      description: 'Billing address',
      isDisabled: false,
    },
    streetAddress1: 'Billing address',
    municipality: 'Hollywood North',
    province: {
      provinceStateId: 1,
      provinceStateCode: 'BC',
      description: 'British Columbia',
    },
    country: {
      countryId: 1,
      countryCode: 'CA',
      description: 'Canada',
    },
    postal: 'V6A 5G7',
  },
  {
    id: 1,
    rowVersion: 3,
    addressType: {
      id: 'MAILING',
      description: 'Mailing address',
      isDisabled: false,
    },
    streetAddress1: 'Mailing address',
    streetAddress2: 'Living in a van',
    streetAddress3: 'Down by the River',
    municipality: 'Hollywood North',
    province: {
      provinceStateId: 1,
      provinceStateCode: 'BC',
      description: 'British Columbia',
    },
    country: {
      countryId: 1,
      countryCode: 'CA',
      description: 'Canada',
    },
    postal: 'V6Z 5G7',
  },
  {
    id: 5,
    rowVersion: 1,
    addressType: {
      id: 'RESIDNT',
      description: 'Property address',
      isDisabled: false,
    },
    streetAddress1: 'Property address',
    streetAddress2: '',
    streetAddress3: '',
    municipality: 'vic',
    province: {
      provinceStateId: 1,
      provinceStateCode: 'BC',
      description: 'British Columbia',
    },
    country: {
      countryId: 1,
      countryCode: 'CA',
      description: 'Canada',
    },
    postal: 'V9A 0H6',
  },
];
