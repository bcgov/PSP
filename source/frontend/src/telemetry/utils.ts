import { metrics, trace } from '@opentelemetry/api';
import { W3CTraceContextPropagator } from '@opentelemetry/core';
import { OTLPMetricExporter } from '@opentelemetry/exporter-metrics-otlp-http';
import { OTLPTraceExporter } from '@opentelemetry/exporter-trace-otlp-http';
import { browserDetector } from '@opentelemetry/opentelemetry-browser-detector';
import { detectResourcesSync, Resource, ResourceAttributes } from '@opentelemetry/resources';
import { MeterProvider, PeriodicExportingMetricReader } from '@opentelemetry/sdk-metrics';
import {
  BatchSpanProcessor,
  ConsoleSpanExporter,
  SimpleSpanProcessor,
  SpanProcessor,
  WebTracerProvider,
} from '@opentelemetry/sdk-trace-web';
import {
  ATTR_SERVICE_NAME,
  ATTR_SERVICE_VERSION,
  ATTR_USER_AGENT_ORIGINAL,
} from '@opentelemetry/semantic-conventions';
import {
  ATTR_DEPLOYMENT_ENVIRONMENT_NAME,
  ATTR_SERVICE_INSTANCE_ID,
} from '@opentelemetry/semantic-conventions/incubating';
import isAbsoluteUrl from 'is-absolute-url';
import { v4 as uuidv4 } from 'uuid';

import { TelemetryConfig } from './config';
import { BrowserAttributesSpanProcessor } from './traces/BrowserAttributesSpanProcessor';
import { UserInfoSpanProcessor } from './traces/UserInfoSpanProcessor';

export const isBrowserEnvironment = () => {
  return typeof window !== 'undefined';
};

export const isNodeEnvironment = () => {
  return typeof process !== 'undefined' && process.release && process.release.name === 'node';
};

// creates URL and appends query parameters
export const buildUrl = (inputUrl: string, queryParams: Record<string, any> = {}): URL => {
  const baseUrl = window.location.origin;
  const urlInstance = isAbsoluteUrl(inputUrl) ? new URL(inputUrl) : new URL(inputUrl, baseUrl);
  Object.keys(queryParams).forEach(k => {
    if (queryParams[k] !== undefined) {
      urlInstance.searchParams.set(k, queryParams[k]);
    }
  });
  return urlInstance;
};

export const isBlocked = (uri: string, config: TelemetryConfig) => {
  const blockList = [...(config.denyUrls ?? []), config.otlpEndpoint];
  return blockList.findIndex(blocked => uri.includes(blocked)) >= 0;
};

// List of meters in the application: e.g. "network", "webvitals", "app", etc
export const NETWORK_METER = 'network-meter';

const makeResource = (config: TelemetryConfig, extraAttributes?: ResourceAttributes) => {
  const uuid = uuidv4();
  let resource = new Resource({
    [ATTR_DEPLOYMENT_ENVIRONMENT_NAME]: config?.environment,
    [ATTR_SERVICE_NAME]: config?.name,
    [ATTR_SERVICE_VERSION]: config?.appVersion,
    [ATTR_SERVICE_INSTANCE_ID]: uuid,
    [ATTR_USER_AGENT_ORIGINAL]: typeof navigator !== 'undefined' ? navigator.userAgent : '',
    'session.instance.id': uuid,
    'screen.width': window.screen.width,
    'screen.height': window.screen.height,
    ...extraAttributes,
  });

  const detectedResources = detectResourcesSync({ detectors: [browserDetector] });
  resource = resource.merge(detectedResources);
  return resource;
};

export const registerMeterProvider = (
  config: TelemetryConfig,
  extraAttributes?: ResourceAttributes,
) => {
  if (config.debug) {
    console.info('[INFO] Registering metrics provider');
  }

  // This is common metadata sent with every metric measurement
  const resource = makeResource(config, extraAttributes);
  const metricExporter = new OTLPMetricExporter({
    url: new URL('/v1/metrics', config.otlpEndpoint).href,
  });

  const meterProvider = new MeterProvider({
    resource: resource,
    readers: [
      new PeriodicExportingMetricReader({
        exporter: metricExporter,
        exportIntervalMillis: config?.exportInterval || 30_000, // export metrics every 30s by default
      }),
    ],
  });

  // set this MeterProvider to be global to the app being instrumented.
  metrics.setGlobalMeterProvider(meterProvider);
};

export const registerTracerProvider = (
  config: TelemetryConfig,
  extraAttributes?: ResourceAttributes,
) => {
  if (config.debug) {
    console.info('[INFO] Registering trace provider');
  }

  // This is common metadata sent with every trace
  const resource = makeResource(config, extraAttributes);
  const exporter = new OTLPTraceExporter({
    url: new URL('v1/traces', config.otlpEndpoint).href,
  });

  const processors: SpanProcessor[] = [];

  if (config.debug) {
    processors.push(new SimpleSpanProcessor(new ConsoleSpanExporter()));
  }

  // use the batch processor for better performance
  processors.push(
    new BatchSpanProcessor(exporter, { scheduledDelayMillis: config?.exportInterval || 5000 }), // export traces every 5s by default
    new BrowserAttributesSpanProcessor(),
    new UserInfoSpanProcessor(),
  );

  const provider = new WebTracerProvider({
    resource,
    spanProcessors: [...processors],
  });

  // set up context propagation
  provider.register({
    propagator: new W3CTraceContextPropagator(),
  });

  // set this TracerProvider to be global to the app
  trace.setGlobalTracerProvider(provider);
};
