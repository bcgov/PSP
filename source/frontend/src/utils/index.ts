export * from './axiosUtils';
export * from './convertUtils';
export * from './mapLookupCode';
export * from './mockKeycloak';
export * from './numberFormatUtils';
export * from './propertyUtils';
export * from './utils';
