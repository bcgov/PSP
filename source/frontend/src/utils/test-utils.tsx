import { useKeycloak } from '@react-keycloak/web';
import {
  fireEvent,
  render as rtlRender,
  RenderOptions as RtlRenderOptions,
  RenderResult,
} from '@testing-library/react';
import { AxiosError, AxiosResponse } from 'axios';
import { FilterProvider } from 'components/maps/providers/FIlterProvider';
import { createMemoryHistory, MemoryHistory } from 'history';
import { IApiError } from 'interfaces/IApiError';
import noop from 'lodash/noop';
import React, { ReactNode } from 'react';
import { MapContainer } from 'react-leaflet';
import { Router } from 'react-router-dom';

import TestCommonWrapper from './TestCommonWrapper';

export const mockKeycloak = (
  props: {
    claims?: string[];
    roles?: string[];
    organizations?: number[];
    authenticated?: boolean;
    userInfo?: any;
  } = {},
) => {
  const { claims, roles, organizations, authenticated = true, userInfo } = props;
  // mock what would be returned by keycloak userinfo endpoint
  const defaultUserInfo = {
    organizations: organizations ?? [1],
    client_roles: [...(claims ?? []), ...(roles ?? [])] ?? [],
    email: 'test@test.com',
    name: 'Chester Tester',
    idir_user_guid: '00000000000000000000000000000000',
  };

  (useKeycloak as jest.Mock).mockReturnValue({
    keycloak: {
      userInfo: userInfo ?? defaultUserInfo,
      subject: 'test',
      authenticated,
      loadUserInfo: jest.fn().mockResolvedValue(userInfo ?? defaultUserInfo),
    },
  });
};

/**
 * Generates fake text data to aid testing input form validation.
 * @param length The length of the text to generate.
 * @returns A string with as many 'x' as the supplied length
 */
export function fakeText(length = 50): string {
  return 'x'.repeat(length);
}

export const fillInput = async (
  container: HTMLElement,
  name: string,
  value: any,
  type: string = 'input',
) => {
  let input: Element | null = null;

  if (type === 'radio') {
    input = container.querySelector(`#input-${name}`);
  } else {
    if (type === 'typeahead' || type === 'datepicker') {
      input = container.querySelector(`input[name="${name}"]`);
    } else {
      input = container.querySelector(`${type}[name="${name}"]`);
    }
  }

  // abort early if no input field found
  if (!input) return { input };
  if (type === 'input') {
    fireEvent.change(input, {
      target: {
        value: value,
      },
    });
    fireEvent.focusOut(input);
  } else if (type === 'typeahead') {
    fireEvent.focus(input);
    fireEvent.change(input, {
      target: {
        value: value,
      },
    });
    const select = container.querySelector(`[aria-label="${value}"]`);
    fireEvent.click(select!);
    fireEvent.focusOut(input);
  } else if (type === 'datepicker') {
    fireEvent.mouseDown(input);
    fireEvent.change(input, {
      target: {
        value: value,
      },
    });
    fireEvent.keyPress(input, { key: 'Enter', code: 'Enter' });
  } else if (type === 'radio') {
    fireEvent.click(input);
    fireEvent.focusOut(input);
  } else {
    fireEvent.change(input, {
      target: {
        value: value,
      },
    });
  }
  fireEvent.blur(input);

  return { input };
};

window.setImmediate = window.setTimeout as any;
export const flushPromises = () => new Promise(window.setImmediate);

export const deferred = () => {
  let resolve: (value?: unknown) => void = noop;
  const promise = new Promise(_resolve => {
    resolve = _resolve;
  });
  return {
    resolve,
    promise,
  };
};

/**
 * Creates an AxiosError with the specified message, config, error code, request and response.
 * @param status The HTTP status code from the server response.
 * @param message The error message.
 * @param data The response that was provided by the server.
 * @returns The created error.
 */
export function createAxiosError(
  status: number = 500,
  message: string = 'Internal Server Error',
  data: any = {},
): AxiosError<IApiError> {
  return {
    isAxiosError: true,
    name: 'AxiosError',
    message,
    config: {},
    toJSON: noop as any,
    response: {
      status,
      data,
    } as AxiosResponse<IApiError>,
  };
}

/**
 * Utility type for generic props of component testing
 */
export interface PropsWithChildren {
  children?: ReactNode;
}

const defaultHistory = createMemoryHistory({
  getUserConfirmation: (message, callback) => callback(true),
});

/**
 * Creates a in-memory router for unit testing
 * @param history (optional) a memory history to use instead of default
 * @returns The route provider
 */
export function createRouteProvider(history?: MemoryHistory) {
  return function Wrapper({ children }: PropsWithChildren) {
    return <Router history={history ?? defaultHistory}>{children}</Router>;
  };
}

/**
 * Creates a Map wrapper for unit testing
 * @param done A callback that will be called when the map has finished rendering
 * @returns The map container instance
 *
 * @example
 *    const { promise, resolve } = deferred();
 *    render(<TestComponent/>, { wrapper: createMapContainer(resolve) })
 *    await waitFor(() => promise)
 *    // the map is fully initialized here...
 */
export function createMapContainer(
  done: () => void = noop,
  whenCreated: (map: L.Map) => void = noop,
) {
  return function Container({ children }: PropsWithChildren) {
    return (
      <div id="mapid" style={{ width: 500, height: 500 }}>
        <MapContainer center={[48.43, -123.37]} zoom={14} whenReady={done} ref={whenCreated}>
          {children}
        </MapContainer>
      </div>
    );
  };
}

// override RenderOptions interface
export interface RenderOptions extends RtlRenderOptions {
  store?: any;
  history?: MemoryHistory;
  useMockAuthentication?: boolean;
  organizations?: number[];
  claims?: string[];
  roles?: string[];
}

function render(
  ui: React.ReactElement,
  options: Omit<RenderOptions, 'wrapper'> = {},
): RenderResult {
  const {
    store,
    history,
    useMockAuthentication = false,
    organizations,
    claims,
    roles,
    ...renderOptions
  } = options;

  // mock authentication state prior to rendering. Check first that keycloak has been mocked!
  if (!!useMockAuthentication || !!claims || !!roles || !!organizations) {
    if (typeof (useKeycloak as jest.Mock).mockReturnValue === 'function') {
      mockKeycloak({
        claims: claims ?? [],
        roles: roles ?? [],
        organizations: organizations ?? [1],
        authenticated: true,
      });
    }
  }

  function AllTheProviders({ children }: PropsWithChildren) {
    return (
      <TestCommonWrapper store={store} history={history}>
        <FilterProvider>{children}</FilterProvider>
      </TestCommonWrapper>
    );
  }
  return rtlRender(ui, { wrapper: AllTheProviders, ...renderOptions });
}

async function renderAsync(
  ui: React.ReactElement,
  options: Omit<RenderOptions, 'wrapper'> = {},
): Promise<RenderResult> {
  const {
    store,
    history,
    useMockAuthentication = false,
    organizations,
    claims,
    roles,
    ...renderOptions
  } = options;

  // mock authentication state prior to rendering. Check first that keycloak has been mocked!
  if (!!useMockAuthentication || !!claims || !!roles || !!organizations) {
    if (typeof (useKeycloak as jest.Mock).mockReturnValue === 'function') {
      mockKeycloak({
        claims: claims ?? [],
        roles: roles ?? [],
        organizations: organizations ?? [1],
        authenticated: true,
      });
    }
  }

  function AllTheProviders({ children }: PropsWithChildren) {
    return (
      <TestCommonWrapper store={store} history={history}>
        <FilterProvider>{children}</FilterProvider>
      </TestCommonWrapper>
    );
  }
  return await rtlRender(ui, { wrapper: AllTheProviders, ...renderOptions });
}

// re-export everything from RTL
export * from '@testing-library/react';
export { default as userEvent } from '@testing-library/user-event';
// override render method
export { render, renderAsync };
