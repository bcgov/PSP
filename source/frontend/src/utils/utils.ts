import { AxiosResponse } from 'axios';
import { FormikProps, getIn } from 'formik';
import { isEmpty, isNull, isUndefined, lowerFirst, startCase } from 'lodash';
import { hideLoading, showLoading } from 'react-redux-loading-bar';

import { SelectOption } from '@/components/common/form';
import { TableSort } from '@/components/Table/TableSort';
import { ApiGen_CodeTypes_DocumentRelationType } from '@/models/api/generated/ApiGen_CodeTypes_DocumentRelationType';
import { EpochIsoDateTime } from '@/models/api/UtcIsoDateTime';
import { logRequest, logSuccess } from '@/store/slices/network/networkSlice';

/**
 * Removes a trailing slash from a string.
 * Useful when creating nested URLs or routes.
 * @param value The input string
 * @returns The string without trailing slash
 */
export function stripTrailingSlash(value: string) {
  return value ? value.replace(/\/$/, '') : value;
}

/**
 * Rounds the supplied number to a certain number of decimal places
 * @param value The number to round
 * @param decimalPlaces The number of decimal places. Defaults to 2
 * @returns The rounded number
 */
export function round(value: number, decimalPlaces = 2): number {
  const factorOfTen = Math.pow(10, decimalPlaces);
  return Math.round((value + Number.EPSILON) * factorOfTen) / factorOfTen;
}

export const enumFromValue = <T extends Record<string, string>>(val: string, _enum: T) => {
  const enumName = (Object.keys(_enum) as Array<keyof T>).find(k => _enum[k] === val);
  if (!enumName) return undefined;
  return _enum[enumName];
};

/**
 * Determine if the specified 'input' value is a positive number of zero.
 * @param input The value to evaluate.
 * @returns True if the value is a positive number or zero, false otherwise.
 */
export const isPositiveNumberOrZero = (input: string | number | undefined | null) => {
  if (isNull(input) || isUndefined(input)) {
    return false;
  }

  if (typeof input === 'string' && isEmpty(input.trim())) {
    return false;
  }

  return !isNaN(Number(input)) && Number(input) > -1;
};

export const isNullOrWhitespace = (value: string | null | undefined): boolean => {
  return !exists(value) || value.trim() === '';
};

type FormikMemoProps = {
  formikProps: FormikProps<any>;
  field: string;
  disabled?: boolean;
  options?: SelectOption[];
  required?: boolean;
} & any;
/**
 * Common use memo function prevents renders unless associated field data has been changed.
 * Essential for performance on large forms. Adapted from:
 * https://jaredpalmer.com/formik/docs/api/fastfield
 * @param param0 params from previous render
 * @param param1 params from current render
 */
export const formikFieldMemo = (
  {
    formikProps: currentProps,
    field: currField,
    disabled: currentDisabled,
    options: currentOptions,
    required: currentRequired,
  }: FormikMemoProps,
  {
    formikProps: prevProps,
    field: prevField,
    disabled: prevDisabled,
    options: prevOptions,
    required: prevRequired,
  }: FormikMemoProps,
) => {
  return !(
    currField !== prevField ||
    currentDisabled !== prevDisabled ||
    currentOptions !== prevOptions ||
    currentRequired !== prevRequired ||
    getIn(currentProps.values, prevField) !== getIn(prevProps.values, prevField) ||
    getIn(currentProps.errors, prevField) !== getIn(prevProps.errors, prevField) ||
    getIn(currentProps.touched, prevField) !== getIn(prevProps.touched, prevField) ||
    Object.keys(currentProps).length !== Object.keys(prevProps).length ||
    currentProps.isSubmitting !== prevProps.isSubmitting
  );
};

/** Provides default redux boilerplate applicable to most axios redux actions
 * @param dispatch Dispatch function
 * @param actionType All dispatched GenericNetworkActions will use this action type.
 * @param axiosPromise The result of an axios.get, .put, ..., call.
 */
export const handleAxiosResponse = <ResponseType>(
  // eslint-disable-next-line @typescript-eslint/ban-types
  dispatch: Function,
  actionType: string,
  axiosPromise: Promise<AxiosResponse<ResponseType>>,
): Promise<AxiosResponse<ResponseType>> => {
  dispatch(logRequest(actionType));
  dispatch(showLoading());
  return axiosPromise
    .then((response: AxiosResponse<ResponseType>) => {
      dispatch(logSuccess({ name: actionType, status: response.status }));
      dispatch(hideLoading());
      return response;
    })
    .finally(() => {
      dispatch(hideLoading());
    });
};

/**
 * convert table sort config to api sort query
 * {name: 'desc} = ['Name desc']
 */
export const generateMultiSortCriteria = (sort: TableSort<any>) => {
  if (!sort) {
    return '';
  }

  return Object.keys(sort).map(key => `${startCase(key).replaceAll(' ', '')} ${sort[key]}`);
};

/**
 * Convert sort query string to TableSort config
 * ['Name desc'] = {name: 'desc'}
 */
export const resolveSortCriteriaFromUrl = (
  input: string[],
): TableSort<any> | Record<string, never> => {
  if (isEmpty(input)) {
    return {};
  }

  return input.reduce((acc: any, sort: string) => {
    const fields = sort.split(' ');
    if (fields.length !== 2) {
      return { ...acc };
    }

    return { ...acc, [lowerFirst(fields[0])]: fields[1] };
  }, {});
};

export function unique<T>(value: T, index: number, array: T[]) {
  return array.indexOf(value) === index;
}

/**
 * Meant to be used as the function passed during a conditional statement to remove null or undefined entries.
 * example. myArray.filter(exists);
 *          if(exists(a?.b?.c)){}
 */
export function exists<T>(value: T | null | undefined): value is T {
  return value === (value ?? !value);
}

/**
 * Returns true id an identifier belongs to an existing entry on the backend
 * @param value the paraneter to be assessed
 * @returns true if valid, false otherwise
 */
export function isValidId(value: number | null | undefined): value is number {
  return exists(value) && !isNaN(value) && value !== 0;
}

export function isValidString(value: string | null | undefined): value is string {
  return exists(value) && value.length > 0;
}

export function isValidIsoDateTime(value: string | null | undefined): value is string {
  return isValidString(value) && value !== EpochIsoDateTime;
}

export function isString(value: unknown): value is string {
  return typeof value === 'string';
}

export function isNumber(value: unknown): value is number {
  return typeof value === 'number';
}

export function truncateName(name: string, length: number): string {
  if (name.length > length) {
    return name.slice(0, length) + '...';
  } else {
    return name;
  }
}

export function firstOrNull<T>(array: Array<T> | null): T | null {
  if (exists(array) && array.length > 0) {
    return array[0];
  }
  return null;
}

/**
 * Add a simple retry wrapper to help avoid chunk errors in deployed pims application, recursively calls promise based on attemptsLeft parameter.
 * @param lazyComponent
 * @param attemptsLeft
 */
export default function componentLoader(lazyComponent: Promise<any>, attemptsLeft: number) {
  return new Promise<any>((resolve, reject) => {
    lazyComponent.then(resolve).catch((error: any) => {
      setTimeout(() => {
        if (attemptsLeft === 0) {
          reject(error);
          return;
        }
        componentLoader(lazyComponent, attemptsLeft - 1).then(resolve, reject);
      }, 500);
    });
  });
}

export function capitalizeFirstLetter(value: string) {
  return value.charAt(0).toUpperCase() + value.slice(1);
}

export function relationshipTypeToPathName(
  relationshipType: ApiGen_CodeTypes_DocumentRelationType,
) {
  switch (relationshipType) {
    case ApiGen_CodeTypes_DocumentRelationType.Templates:
      return 'template';

    case ApiGen_CodeTypes_DocumentRelationType.ResearchFiles:
      return 'research';

    case ApiGen_CodeTypes_DocumentRelationType.AcquisitionFiles:
      return 'acquisition';

    case ApiGen_CodeTypes_DocumentRelationType.Leases:
      return 'lease';

    case ApiGen_CodeTypes_DocumentRelationType.Projects:
      return 'project';

    case ApiGen_CodeTypes_DocumentRelationType.ManagementActivities:
      return 'activities';

    case ApiGen_CodeTypes_DocumentRelationType.ManagementFiles:
      return 'management';

    case ApiGen_CodeTypes_DocumentRelationType.DispositionFiles:
      return 'disposition';

    case ApiGen_CodeTypes_DocumentRelationType.Properties:
      return 'property';
  }
}
