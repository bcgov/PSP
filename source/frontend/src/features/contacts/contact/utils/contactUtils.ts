import { CountryCodes } from 'constants/index';
import { AddressField } from 'features/contacts/interfaces';
import { validateYupSchema, yupToFormErrors } from 'formik';
import { IEditableOrganizationForm, IEditablePersonForm } from 'interfaces/editable-contact';
import isEmpty from 'lodash/isEmpty';

import {
  hasAddress,
  hasEmail,
  hasPhoneNumber,
  OrganizationValidationSchema,
  PersonValidationSchema,
} from '../create/validation';
import { sortAddresses } from '../detail/utils';
import { IContactAddress } from './../../../../interfaces/IContact';

export const onValidateOrganization = (
  values: IEditableOrganizationForm,
  otherCountryId?: string,
) => {
  const errors = {} as any;
  try {
    // combine yup schema validation with custom rules
    if (!hasEmail(values) && !hasPhoneNumber(values) && !hasAddress(values)) {
      errors.needsContactMethod =
        'Contacts must have a minimum of one method of contact to be saved. (ex: email,phone or address)';
    }
    validateYupSchema(values, OrganizationValidationSchema, true, {
      otherCountry: otherCountryId,
    });
    return errors;
  } catch (err) {
    return { ...errors, ...yupToFormErrors(err) };
  }
};

export const onValidatePerson = (values: IEditablePersonForm, otherCountryId?: string) => {
  const errors = {} as any;
  try {
    // combine yup schema validation with custom rules
    if (!hasEmail(values) && !hasPhoneNumber(values) && !hasAddress(values)) {
      errors.needsContactMethod =
        'Contacts must have a minimum of one method of contact to be saved. (ex: email,phone or address)';
    }
    validateYupSchema(values, PersonValidationSchema, true, { otherCountry: otherCountryId });

    return errors;
  } catch (err) {
    return { ...errors, ...yupToFormErrors(err) };
  }
};

export const toAddressFields = (addresses: IContactAddress[]) => {
  return addresses
    .sort(sortAddresses)
    .reduce((accumulator: AddressField[], value: IContactAddress) => {
      accumulator.push({
        label: value.addressType.description || '',
        streetAddress1: value.streetAddress1,
        streetAddress2: value.streetAddress2,
        streetAddress3: value.streetAddress3,
        municipalityAndProvince: [
          value.municipality !== undefined ? value.municipality : '',
          value.province?.provinceStateCode ?? '',
        ]
          .filter(x => !isEmpty(x))
          .join(' '),
        country:
          value.country?.countryCode === CountryCodes.Other
            ? value.countryOther ?? ''
            : value.country?.description,
        postal: value.postal,
      });
      return accumulator;
    }, []);
};
