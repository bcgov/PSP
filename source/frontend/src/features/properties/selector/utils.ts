import { IMapProperty } from 'features/properties/selector/models';
import { compact } from 'lodash';
import { Api_PropertyFile } from 'models/api/PropertyFile';
import { formatApiAddress, pidFormatter } from 'utils';

import { Api_Geometry, Api_Property } from './../../../models/api/Property';

export enum NameSourceType {
  PID = 'PID',
  PIN = 'PIN',
  PLAN = 'Plan #',
  LOCATION = 'Location',
  NAME = 'Descriptive Name',
  NONE = 'n/a',
  ADDRESS = 'Address',
}

interface PropertyName {
  label: NameSourceType;
  value: string;
}

export const getPropertyName = (property: IMapProperty): PropertyName => {
  if (property.pid !== undefined && property.pid?.toString().length > 0 && property.pid !== '0') {
    return { label: NameSourceType.PID, value: pidFormatter(property.pid.toString()) };
  } else if (
    property.pin !== undefined &&
    property.pin?.toString()?.length > 0 &&
    property.pin !== '0'
  ) {
    return { label: NameSourceType.PIN, value: property.pin.toString() };
  } else if (property.planNumber !== undefined) {
    return { label: NameSourceType.PLAN, value: property.planNumber };
  } else if (property.latitude !== undefined && property.longitude !== undefined) {
    return {
      label: NameSourceType.LOCATION,
      value: compact([property.longitude?.toFixed(6), property.latitude?.toFixed(6)]).join(', '),
    };
  } else if (property.address !== undefined) {
    return {
      label: NameSourceType.ADDRESS,
      value: property.address,
    };
  }
  return { label: NameSourceType.NONE, value: '' };
};

export const getPrettyLatLng = (location?: Api_Geometry) =>
  compact([
    location?.coordinate?.x?.toFixed(6) ?? 0,
    location?.coordinate?.y?.toFixed(6) ?? 0,
  ]).join(', ');

export const getFilePropertyName = (
  fileProperty?: Api_PropertyFile,
  skipName: boolean = false,
): PropertyName => {
  if (fileProperty === undefined) {
    return { label: NameSourceType.NONE, value: '' };
  }

  if (
    fileProperty.propertyName !== undefined &&
    fileProperty.propertyName !== '' &&
    skipName === false
  ) {
    return { label: NameSourceType.NAME, value: fileProperty.propertyName };
  } else if (fileProperty.property !== undefined) {
    const property = fileProperty.property;
    return getApiPropertyName(property);
  }
  return { label: NameSourceType.NONE, value: '' };
};

export const getApiPropertyName = (property?: Api_Property): PropertyName => {
  if (property === undefined) {
    return { label: NameSourceType.NONE, value: '' };
  }

  let mapProperty: IMapProperty = {
    pin: property.pin?.toString(),
    pid: property.pid?.toString(),
    latitude: property.latitude,
    longitude: property.longitude,
    planNumber: property.planNumber,
    address: property.address !== undefined ? formatApiAddress(property.address) : undefined,
  };
  return getPropertyName(mapProperty);
};
