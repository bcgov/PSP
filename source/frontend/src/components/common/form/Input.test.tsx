import { Formik } from 'formik';
import noop from 'lodash/noop';
import { act, render, RenderOptions, userEvent, waitFor } from 'utils/test-utils';

import { Input, InputProps } from './';

const BASIC_PROPS: InputProps = { field: 'foo' };
const FORMIK_VALUES = { foo: '' };

describe('Input component', () => {
  // render component under test
  const setup = (
    props: InputProps = BASIC_PROPS,
    initialValues: { foo?: string } = FORMIK_VALUES,
    renderOptions: RenderOptions = {},
  ) => {
    const { label = 'Test Label', ...rest } = props;
    const utils = render(
      <Formik initialValues={initialValues ?? {}} onSubmit={noop}>
        <Input label={label} {...rest} />
      </Formik>,
      {
        ...renderOptions,
      },
    );

    return { ...utils };
  };

  beforeEach(() => {});

  afterEach(() => {
    jest.resetAllMocks();
  });

  it('renders as expected', () => {
    const { asFragment } = setup();
    expect(asFragment()).toMatchSnapshot();
  });

  it('displays existing values if they exist', () => {
    const { getByRole } = setup(BASIC_PROPS, { foo: 'bar' });
    const input = getByRole('textbox');

    expect(input).toBeVisible();
    expect(input).toHaveValue('bar');
    expect(input.tagName).toBe('INPUT');
  });

  it('supports disabled property as intended', () => {
    const { getByRole } = setup({ ...BASIC_PROPS, disabled: true }, { foo: 'bar' });
    const input = getByRole('textbox');

    expect(input).toBeVisible();
    expect(input).toHaveValue('bar');
    expect(input.tagName).toBe('INPUT');
    expect(input).toHaveAttribute('disabled');
  });

  it('displays associated help text if provided', () => {
    const help = 'This field is case-sensitive';
    const { getByText } = setup({ ...BASIC_PROPS, helpText: help });

    expect(getByText(help)).toBeVisible();
  });

  it(`renders as TEXTAREA when property 'as' equals 'textarea'`, () => {
    const { getByRole } = setup({ ...BASIC_PROPS, as: 'textarea' }, { foo: 'bar' });
    const input = getByRole('textbox');

    expect(input).toBeVisible();
    expect(input).toHaveValue('bar');
    expect(input.tagName).toBe('TEXTAREA');
  });

  it('calls onChange callback when value changes', async () => {
    const onChangeSpy = jest.fn();
    const { getByRole } = setup({ ...BASIC_PROPS, onChange: onChangeSpy });

    // type some random data into text input
    await act(async () => userEvent.type(getByRole('textbox'), 'Lorem ipsum'));

    // assert
    await waitFor(() => expect(onChangeSpy).toHaveBeenCalled());
  });
});
