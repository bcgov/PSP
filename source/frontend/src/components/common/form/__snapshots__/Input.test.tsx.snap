// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Input component renders as expected 1`] = `
<DocumentFragment>
  <div
    class="Toastify"
  />
  <div
    class="container"
  />
  <div
    class="input form-group"
  >
    <label
      class="form-label"
      for="input-foo"
    >
      Test Label 
    </label>
    <input
      class="form-control"
      id="input-foo"
      name="foo"
      title=""
      value=""
    />
  </div>
</DocumentFragment>
`;
